src/main/java/org/sagebionetworks/bridge/services/AdherenceService.java

```diff
@@ -376,7 +378,70 @@ public void deleteAdherenceRecord(AdherenceRecord record) {
             recordDao.deleteAdherenceRecordPermanently(record);
         }
     }
-    
+
+    /** Update just the specific post-processing attributes on the adherence record. */
+    public void updateAdherencePostProcessingAttributes(String appId, String studyId, String userId,
+            String instanceGuid, DateTime eventTimestamp, AdherencePostProcessingAttributes attributes) {
+        checkNotNull(appId);
+        checkNotNull(studyId);
+        checkNotNull(userId);
+        checkNotNull(instanceGuid);
+        checkNotNull(eventTimestamp);
+        checkNotNull(attributes);
+
+        // Check permissions.
+        CAN_ACCESS_ADHERENCE_DATA.checkAndThrow(
+                AuthEvaluatorField.STUDY_ID, studyId,
+                AuthEvaluatorField.USER_ID, userId
+        );
+
+        // Get the adherence record that we need to update.
+        AdherenceRecordsSearch search = new AdherenceRecordsSearch.Builder()
+                .withStudyId(studyId)
+                .withUserId(userId)
+                .withInstanceGuids(ImmutableSet.of(instanceGuid))
+
+                // Event timestamp is a range of values. Start is inclusive, end is exclusive. Since we're looking for
+                // a specific singular timestamp, the end is start+1.
+                .withEventTimestampStart(eventTimestamp)
+                .withEventTimestampEnd(eventTimestamp.plusMillis(1))
+
+                // Page size is required, but we're only looking for one record, so this is fine.
+                .withPageSize(1)
+                .build();
+        List<AdherenceRecord> adherenceRecordList = recordDao.getAdherenceRecords(search).getItems();
+
+        AdherenceRecord record;
+        if (adherenceRecordList.isEmpty()) {
+            // If there's no record, create one.
+            record = new AdherenceRecord();
+            record.setAppId(appId);
+            record.setStudyId(studyId);
+            record.setUserId(userId);
+            record.setInstanceGuid(instanceGuid);
+            record.setEventTimestamp(eventTimestamp);
+        } else {
+            // Otherwise, update the existing record. There should only be one record.
+            record = adherenceRecordList.get(0);
+        }
+
+        // Update the record.
+        record.setPostProcessingAttributes(attributes.getPostProcessingAttributes());
+        record.setPostProcessingCompletedOn(attributes.getPostProcessingCompletedOn());
+        record.setPostProcessingStatus(attributes.getPostProcessingStatus());
```

Comment:
Only write these if they are non-null. Merge old attributes with new attributes.

Reply:
Put this logic in service.updateAdherenceRecords() so that no one can accidentally clobber these attributes.